cuda_core/cuda/core/experimental/_memory/_memory_pool.pyx

```diff
@@ -0,0 +1,438 @@
+            ELSE:
+                raise RuntimeError("not supported")
+        else:
+            IF CUDA_CORE_BUILD_MAJOR >= 13:
+                if opts._type == cydriver.CUmemAllocationType.CU_MEM_ALLOCATION_TYPE_MANAGED:
+                    # Managed memory pools
+                    loc.id = dev_id
+                    loc.type = opts._location
+                    HANDLE_RETURN(cydriver.cuMemGetMemPool(&(self._handle), &loc, opts._type))
+                else:
+                    assert False
+            ELSE:
+                assert False
+
+    return 0
+
+
+cdef int _MP_init_create(_MemPool self, int dev_id, _MemPoolOptions opts) except?-1:
+    cdef cydriver.CUmemPoolProps properties
+    memset(&properties, 0, sizeof(cydriver.CUmemPoolProps))
+
+    cdef bint ipc_enabled = opts._ipc_enabled
+    properties.allocType = opts._type
+    properties.handleTypes = _ipc.IPC_HANDLE_TYPE if ipc_enabled else cydriver.CUmemAllocationHandleType.CU_MEM_HANDLE_TYPE_NONE
+    properties.location.id = dev_id
+    properties.location.type = opts._location
+    # managed memory does not support maxSize as of CUDA 13.0
+    IF CUDA_CORE_BUILD_MAJOR >= 13:
+        if properties.allocType != cydriver.CUmemAllocationType.CU_MEM_ALLOCATION_TYPE_MANAGED:
+            properties.maxSize = opts._max_size
+    ELSE:
+        properties.maxSize = opts._max_size
+
+    self._dev_id = dev_id
+    self._mempool_owned = True
+
+    with nogil:
+        HANDLE_RETURN(cydriver.cuMemPoolCreate(&(self._handle), &properties))
+        # TODO: should we also set the threshold here?
+
+    if ipc_enabled:
+        alloc_handle = _ipc.MP_export_mempool(self)
+        self._ipc_data = _ipc.IPCDataForMR(alloc_handle, False)
+
+    return 0
+
+
+# Raise an exception if the given stream is capturing.
+# A result of CU_STREAM_CAPTURE_STATUS_INVALIDATED is considered an error.
+cdef inline int check_not_capturing(cydriver.CUstream s) except?-1 nogil:
+    cdef cydriver.CUstreamCaptureStatus capturing
+    HANDLE_RETURN(cydriver.cuStreamIsCapturing(s, &capturing))
+    if capturing != cydriver.CUstreamCaptureStatus.CU_STREAM_CAPTURE_STATUS_NONE:
+        raise RuntimeError("_MemPool cannot perform memory operations on "
+                           "a capturing stream (consider using GraphMemoryResource).")
+
+
+cdef inline Buffer _MP_allocate(_MemPool self, size_t size, Stream stream):
+    cdef cydriver.CUstream s = stream._handle
+    cdef cydriver.CUdeviceptr devptr
+    with nogil:
+        check_not_capturing(s)
+        HANDLE_RETURN(cydriver.cuMemAllocFromPoolAsync(&devptr, size, self._handle, s))
+    cdef Buffer buf = Buffer.__new__(Buffer)
+    buf._ptr = <uintptr_t>(devptr)
+    buf._ptr_obj = None
+    buf._size = size
+    buf._memory_resource = self
+    buf._alloc_stream = stream
+    return buf
+
+
+cdef inline void _MP_deallocate(
+    _MemPool self, uintptr_t ptr, size_t size, Stream stream
+) noexcept:
```

Comment:
Can we mark this function as `nogil`? I don't see anywhere needing the GIL here?

Reply:
It turns out that this is an optimal implementation. We creates the C args from the Python objects at the call site, so even if we mark this as nogil, we can't possibly relax the gil at the call site anyway.